### ISSUE TITLE
SchemaDiffTool: schema sync script error when changing output data type or OUT argument list in function

### ISSUE THREAD

**Issue by firstDismay** [pgadmin-org/pgadmin4#6652]
Changing the output data type or OUT argument list in the source schema function causes the target schema synchronization script to fail. This change is not considered an overload in PostgreSQL and cannot be applied until the function is removed.
Problem reproduction:
Scheme source:
```
CREATE OR REPLACE FUNCTION class_del(
	iid bigint)
    RETURNS void
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE SECURITY DEFINER PARALLEL UNSAFE
AS $BODY$
    --source FUNCTION new API 
DECLARE
    action_state character varying DEFAULT 'none';
BEGIN
    ---Code Function
	--************************************************************
	----------------THROW EXCEPTION
	--************************************************************
	IF (action_state <> 'action_allowed') THEN
		RAISE EXCEPTION 'Error message';
	END IF;	
END;
$BODY$;
```
Scheme target:
```
CREATE OR REPLACE FUNCTION class_del(
	iid bigint,
	OUT Result INT)
    RETURNS INT
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE SECURITY DEFINER PARALLEL UNSAFE
AS $BODY$
    --target FUNCTION old API 
DECLARE
    action_state character varying DEFAULT 'none';
BEGIN
	Result = 0;
    ---Code Function
	--************************************************************
	----------------THROW EXCEPTION
	--************************************************************
	IF (action_state <> 'action_allowed') THEN
		Result = 144; --Set error code
	END IF;	
END;
$BODY$;
```

**Comment by akshay-joshi** [comment]
@firstDismay 

If I understood correctly I have run the above-given query and then run the schema diff which is showing two objects, one is source only and another is target only because of function overloading.
<img width="1336" alt="Screenshot 2023-08-14 at 1 36 17 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/1632882/fae15426-233e-40e7-8013-c5caeea6cf3b">

The above behavior is expected because for schema diff both the functions are different. Run the below test case:
Source function:
```
CREATE OR REPLACE FUNCTION source.test_sum(a bigint, b bigint)
    RETURNS bigint
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE PARALLEL UNSAFE
AS $BODY$
BEGIN
RETURN a + b;
END;
$BODY$; 
``` 

Target Function:
```
CREATE OR REPLACE FUNCTION target.test_sum(a bigint, b bigint)
    RETURNS double precision
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE PARALLEL UNSAFE
AS $BODY$
BEGIN
RETURN a + b;
END;
$BODY$;
```

In short, if the function's return type is different than schema diff will add the 'DROP FUNCTION..' syntax.

**Comment by firstDismay** [comment]
I agree with this behavior. But drop will be added after create. In case of overload, this sequence is erroneous. The script will not run until you swap drop and create for my example.

**Comment by akshay-joshi** [comment]
@firstDismay 

I am unable to reproduce that, in my case DROP statement is before the CREATE.
<img width="1301" alt="Screenshot 2023-08-16 at 12 31 39 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/1632882/b417e0fe-7437-400a-be89-e93f0962e63e">

**Comment by firstDismay** [comment]
@akshay-joshi 
I didn't use this form. I clicked the create script button for all the differences in the functions. For version 7.4, I had all drop function operations at the end. I swapped places manually. After that, everything was successful. I don't have access to the computer now I can't check your option. I have more than 700 functions, one does not suit me.

**Comment by akshay-joshi** [comment]
@firstDismay

I am still not able to reproduce. I have created 7 identical functions with different names and return types in source and target. I selected all of them and click on generate script button.
<img width="1314" alt="Screenshot 2023-08-16 at 1 11 33 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/1632882/b9e65b25-0aca-4bbf-9afc-a74ca9308dc7">
<img width="1314" alt="Screenshot 2023-08-16 at 1 11 50 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/1632882/7ae107c3-6043-4aac-8ddf-3a4aa6cacbff">

**Comment by firstDismay** [comment]
@akshay-joshi 
I have not observed this behavior. I wrote about this problem earlier via the e-mail mailing list. Version 7.5 came out right after that. I had no reason to think that the problem was given attention in this issue. I will get access to the computer after 20.08 and check again in the current version.

**Comment by firstDismay** [comment]
I don't see the behavior you describe when adding a DROP is needed. I changed the output type of the function from void to record. Added two OUT variables. Conducted a comparison and received a synchronization script:

**Comment by akshay-joshi** [comment]
@firstDismay

If I understand correctly schema diff is showing two objects one is **source only** and another is **target only** for functions "**class_prop_del**" and "**class_del**" just because their signature is different(Function overloading). When you select them it will create the "CREATE OR REPLACE" script and then the "DROP" statement.

If the above is correct then it's been difficult to fix the issue, as schema diff treated both the functions as separate functions so on your selection it will create the statement. I suggest you apply the 'Target Only' filter which will create the 'DROP syntax' for all 700 functions. Select all of them and click on the "Generate script" button. 

Then apply filter 'Source Only', select all the functions, and click on the "Generate script" button.  
  
<img width="254" alt="Screenshot 2023-08-25 at 1 40 26 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/1632882/2a911f13-82e0-4b26-9d45-6e1729e67774">

**Comment by firstDismay** [comment]
@akshay-joshi
From the point of view of PostgreSQL, these functions are equivalent. The signature of the function is determined only by the arguments of the IN, INOUT and VARIADIC. 
[proargtypes](https://www.postgresql.org/docs/current/catalog-pg-proc.html) - An array of the data types of the function arguments. This includes only input arguments (including INOUT and VARIADIC arguments), and thus represents the call signature of the function.
Based on these considerations, you can install DROP before CREATE.

**Comment by akshay-joshi** [comment]
@firstDismay 

This is very complex for Schema Diff to compare considering all the scenarios. Schema Diff compares the functions with Name and Argument. I am keeping it as a Low Priority issue.

**Comment by firstDismay** [comment]
@akshay-joshi
I do not insist on the seriousness of the problem. I'm just pointing out that your interpretation of overloading doesn't follow PostgreSQL's call signature principles, which is the basis for the divergence in defining function equivalence, which leads to separate processing when signatures differ in areas not considered by PostgreSQL.
